File: app/src/androidTest/java/com/backbase/citysearch/TrieImplTest.java
package com.backbase.citysearch;

import com.backbase.citysearch.models.City;
import com.backbase.citysearch.searchalgo.Index;
import com.backbase.citysearch.searchalgo.TrieImpl;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

@RunWith(JUnit4.class)
public class TrieImplTest {
    @Test
    public void searchWithCorrectQueryShouldReturnSearchResult() {
        Index<City> index = new TrieImpl<>();
        String cityName = "Sydney";
        String countryName = "AU";
        City testCity = new City();
        testCity.setCountry(countryName);
        testCity.setName(cityName);
        index.addEntry(cityName, testCity);
        List<City> searchCityResult = index.findEntries(cityName);
        assertEquals(1, searchCityResult.size());
        assertEquals(testCity.getName(), searchCityResult.get(0).getName());
        assertEquals(testCity.getCountry(), searchCityResult.get(0).getCountry());
    }

    @Test
    public void searchWithWrongQueryShouldReturnNull() {
        Index<City> index = new TrieImpl<>();
        List<City> searchCityResult = index.findEntries("Sydney");
        assertNull(searchCityResult);
    }
}

File: app/src/androidTest/java/com/backbase/citysearch/TrieImplTest.kt
package com.backbase.citysearch

import com.backbase.citysearch.models.City
import com.backbase.citysearch.searchalgo.Index
import com.backbase.citysearch.searchalgo.TrieImpl
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class TrieImplTest {
    @Test
    fun searchWithCorrectQueryShouldReturnSearchResult() {
        val index: Index<City> = TrieImpl()
        val cityName = "Sydney"
        val countryName = "AU"
        val testCity = City()
        testCity.country = countryName
        testCity.name = cityName
        index.addEntry(cityName, testCity)
        val searchCityResult = index.findEntries(cityName)
        Assert.assertEquals(1, searchCityResult!!.size.toLong())
        Assert.assertEquals(testCity.name, searchCityResult[0].name)
        Assert.assertEquals(testCity.country, searchCityResult[0].country)
    }

    @Test
    fun searchWithWrongQueryShouldReturnNull() {
        val index: Index<City> = TrieImpl()
        val searchCityResult = index.findEntries("Sydney")
        Assert.assertNull(searchCityResult)
    }
}